src/Hosting/TestHost/src/ResponseFeature.cs

```diff
@@ -98,14 +99,22 @@ public void OnCompleted(Func<object, Task> callback, object state)
 
         public async Task FireOnSendingHeadersAsync()
         {
-            await _responseStartingAsync();
-            HasStarted = true;
-            _headers.IsReadOnly = true;
+            if (!HasStarted)
+            {
+                await _responseStartingAsync();
+                HasStarted = true;
+                _headers.IsReadOnly = true;
+            }
         }
 
         public Task FireOnResponseCompletedAsync()
         {
             return _responseCompletedAsync();
         }
+
+        public Task StartAsync(CancellationToken token = default)
+        {
+            return FireOnSendingHeadersAsync();
```

Comment:
If this throws then HasStarted doesn't get set and FireOnSendingHeadersAsync can be called again.

Reply:
Done.

Is it correct that the response always gets returned when an error is thrown during StartAsync. I needed to do that to get `ExceptionFromOnStartingWithErrorHandlerIsReported` to pass. The test isn't doing what it says it is.

https://github.com/aspnet/AspNetCore/blob/11311d4f9d7eafcbe2ebafa7a5e8126244187a57/src/Hosting/TestHost/test/ClientHandlerTests.cs#L357-L380